enhancements/update/clusterversion-history-pruning.md

```diff
@@ -0,0 +1,166 @@
+  - 20 * (isPartialWithinAZStream)
+  - 1.01 * sliceIndex
+```
+The lowest ranked entry will be removed, or, in the case of the new entry not added, maintaining the queue size of 100 entries.
+The removal including details of why an entry was chosen for removal will be logged.
+
+The weights used to rank a given entry are based on the following criteria:
+* `1000` - considered most important to keep
+* `30` - interesting but not critical  
+* `20` - partial stops during a minor transition are interesting, but not as interesting as the completed versions that bookend the minor transition
+* `-20` - partials between completed releases within the same minor, not very useful
+* `-1.01` - prefer more recent entries, avoid ties
+
+`isAFinalEntry` will be true for the 5 most recent history entries.
+
+Prototyping against ClusterVersion histories from a variety of actual customer clusters will help determine the correct criteria and corresponding weights that result in the most useful history.
+
+godocs will be created explaining that ClusterVersion history may be missing entries (has been pruned).
+
+### API Extensions
+
+N/A - no api extensions are being introduced in this proposal
+
+### Risks and Mitigations
+
+* Useful ClusterVersion history could be removed.
+This is also a risk if we do nothing.
+By defining a pruning algorithm we at least have some control over what is removed.
+
+### Drawbacks
+
+* It is possible that the current pruning method, which simply retains the most recent 50 ClusterVersion history entries, would result in an overall history that is more useful for identifying and resolving a given issue.
+However in the vast majority of instances this is not expected.
+
+## Design Details
+
+### Test Plan
+
+* CVO unit tests will be expanded and/or created to test the new logic.
+
+### Graduation Criteria
+
+GA. When it works, we ship it.
+
+#### Dev Preview -> Tech Preview
+
+N/A. This is not expected to be released as Dev Preview.
+
+#### Tech Preview -> GA
+
+N/A. This is not expected to be released as Tech Preview.
+
+#### Removing a deprecated feature
+
+N/A.
+
+### Upgrade / Downgrade Strategy
+
+No special consideration.
+
+### Version Skew Strategy
+
+No special consideration.
+
+### Operational Aspects of API Extensions
+
+N/A - no api extensions are being introduced in this proposal
+
+#### Failure Modes
+
+No special consideration.
+
+#### Support Procedures
+
+No special consideration.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation History`.
+
+## Alternatives
+
+### Significantly enlarge ClusterVersion history queue
+
+The ClusterVersion history queue could be enlarged to a point where it is hoped that for a majority of clusters, all, or a significant amount, of the cluster's history can be retained.
+
+* No in-depth analysis has been done to determine the required queue size.
+* This would increase memory usage.
```

Comment:
there's also an etcd/Kube cap on resource size, so at some point this approach would result in old clusters failing to update `ClusterVersion` because the rendered object exceeded that size.

Reply:
added